# CMOR 350 HW 7 Question 2 Part B
The goal is to find the occupancy times of this rate matrix, in order to calculate the amount of time Machine 1 is blocked in the first hour of operation.

Here's the rate matrix.

In [102]:
mu1 = 1/1
mu2 = 1/1.2
mu3 = 1/1

r_matrix = [
    #1,     2,       3,      4,     5,      6,      7,      #8
    [0,     mu1,     0,     0,      0,      0,      0,      0],     #1
    [0,     0,       mu1,   0,      0,      mu2,    0,      0],     #2
    [0,     0,       0,     mu2,    0,      0,      0,      0],     #3 B,W,I
    [0,     0,       mu3,   0,      mu2,    0,      0,      mu1],   #4
    [0,     0,       0,     0,      0,      mu3,    mu1,    0],     #5 
    [mu3,   0,       0,     mu1,    0,      0,      0,      0],     #6 W,I,W
    [0,     0,       0,     mu3,    0,      0,      0,      0],     #7 B,B,W
    [0,     0,       mu3,   0,      0,      0,      mu2,    0],     #8 B,W,W
]
r_matrix

[[0, 1.0, 0, 0, 0, 0, 0, 0],
 [0, 0, 1.0, 0, 0, 0.8333333333333334, 0, 0],
 [0, 0, 0, 0.8333333333333334, 0, 0, 0, 0],
 [0, 0, 1.0, 0, 0.8333333333333334, 0, 0, 1.0],
 [0, 0, 0, 0, 0, 1.0, 1.0, 0],
 [1.0, 0, 0, 1.0, 0, 0, 0, 0],
 [0, 0, 0, 1.0, 0, 0, 0, 0],
 [0, 0, 1.0, 0, 0, 0, 0.8333333333333334, 0]]

In [103]:
for row in r_matrix:
    print(sum(row))

1.0
1.8333333333333335
0.8333333333333334
2.8333333333333335
2.0
2.0
1.0
1.8333333333333335


In [104]:
row_sum = []
for row in r_matrix:
    row_sum.append(sum(row))
nu = max(row_sum)
row_sum, nu

([1.0,
  1.8333333333333335,
  0.8333333333333334,
  2.8333333333333335,
  2.0,
  2.0,
  1.0,
  1.8333333333333335],
 2.8333333333333335)

In [105]:
p_hat = []
for row in r_matrix:
    p_hat.append([x / nu for x in row])
p_hat

[[0.0, 0.3529411764705882, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
 [0.0, 0.0, 0.3529411764705882, 0.0, 0.0, 0.29411764705882354, 0.0, 0.0],
 [0.0, 0.0, 0.0, 0.29411764705882354, 0.0, 0.0, 0.0, 0.0],
 [0.0,
  0.0,
  0.3529411764705882,
  0.0,
  0.29411764705882354,
  0.0,
  0.0,
  0.3529411764705882],
 [0.0, 0.0, 0.0, 0.0, 0.0, 0.3529411764705882, 0.3529411764705882, 0.0],
 [0.3529411764705882, 0.0, 0.0, 0.3529411764705882, 0.0, 0.0, 0.0, 0.0],
 [0.0, 0.0, 0.0, 0.3529411764705882, 0.0, 0.0, 0.0, 0.0],
 [0.0, 0.0, 0.3529411764705882, 0.0, 0.0, 0.0, 0.29411764705882354, 0.0]]

In [106]:
for row in p_hat:
    print(sum(row))

0.3529411764705882
0.6470588235294117
0.29411764705882354
0.9999999999999999
0.7058823529411764
0.7058823529411764
0.3529411764705882
0.6470588235294117


In [107]:
for i in range(len(p_hat)):
    p_hat[i][i] = 1 - sum(p_hat[i])
p_hat

[[0.6470588235294118, 0.3529411764705882, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
 [0.0,
  0.3529411764705883,
  0.3529411764705882,
  0.0,
  0.0,
  0.29411764705882354,
  0.0,
  0.0],
 [0.0, 0.0, 0.7058823529411764, 0.29411764705882354, 0.0, 0.0, 0.0, 0.0],
 [0.0,
  0.0,
  0.3529411764705882,
  1.1102230246251565e-16,
  0.29411764705882354,
  0.0,
  0.0,
  0.3529411764705882],
 [0.0,
  0.0,
  0.0,
  0.0,
  0.2941176470588236,
  0.3529411764705882,
  0.3529411764705882,
  0.0],
 [0.3529411764705882,
  0.0,
  0.0,
  0.3529411764705882,
  0.0,
  0.2941176470588236,
  0.0,
  0.0],
 [0.0, 0.0, 0.0, 0.3529411764705882, 0.0, 0.0, 0.6470588235294118, 0.0],
 [0.0,
  0.0,
  0.3529411764705882,
  0.0,
  0.0,
  0.0,
  0.29411764705882354,
  0.3529411764705883]]

In [108]:
for row in p_hat:
    print(sum(row))

1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0


### $\hat P$ Is Now Created

Now we want 
$$\begin{align}
M(T)&=\sum_{k=0}^\infin\frac{1}{\nu}P(Y>k)\hat{P}^k\\
&=\sum_{k=0}^\infin\frac{1}{\nu}(1-\sum_{l=0}^ke^{-\nu T}\frac{(\nu T)^l}{l!}))\hat{P}^k
\end{align}$$
To get the Occupancy matrix for the first hour of operations.

Note - $\infin$ is too large, so we can approximate it with 
$$max(\nu t + 5\sqrt{\nu t},20)$$

$$M(T) = \sum_{k=0}^{max(\nu t + 5\sqrt{\nu t},20)}\frac{1}{\nu}(1-\sum_{l=0}^ke^{-\nu T}\frac{(\nu T)^l}{l!}))\hat{P}^k$$


### Note; units
Our $\mu_i$ are given in terms of minutes, but we are asked about the first hour. Thus, we'll be sure to look at `M(60)`



In [109]:
import numpy as np

p_hat=np.array(p_hat)

p_hat

array([[6.47058824e-01, 3.52941176e-01, 0.00000000e+00, 0.00000000e+00,
        0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00],
       [0.00000000e+00, 3.52941176e-01, 3.52941176e-01, 0.00000000e+00,
        0.00000000e+00, 2.94117647e-01, 0.00000000e+00, 0.00000000e+00],
       [0.00000000e+00, 0.00000000e+00, 7.05882353e-01, 2.94117647e-01,
        0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00],
       [0.00000000e+00, 0.00000000e+00, 3.52941176e-01, 1.11022302e-16,
        2.94117647e-01, 0.00000000e+00, 0.00000000e+00, 3.52941176e-01],
       [0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
        2.94117647e-01, 3.52941176e-01, 3.52941176e-01, 0.00000000e+00],
       [3.52941176e-01, 0.00000000e+00, 0.00000000e+00, 3.52941176e-01,
        0.00000000e+00, 2.94117647e-01, 0.00000000e+00, 0.00000000e+00],
       [0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 3.52941176e-01,
        0.00000000e+00, 0.00000000e+00, 6.47058824e-01, 0.

In [110]:
import math

def calculate_M(T, nu, P_hat):
    max_k = round(max(nu*T + 5*math.sqrt(nu*T),20))
    print(max_k)
    total = 0
    for k in range(max_k):
        inner_sum = sum(math.exp(-nu * T) * ((nu * T) ** l) / math.factorial(l) for l in range(k + 1))
        total += (1 / nu) * (1 - inner_sum) * np.linalg.matrix_power(P_hat, k)
    return total

# Distribution of the first 60 min
occupancy_matrix_60min = calculate_M(60, nu, p_hat)
occupancy_matrix_60min

235


OverflowError: (34, 'Result too large')

### Errors in this formulation
I got `overflow` errors when I exceeded 140 iterations of the outerloop. e.g., $k\ge140$ so I had to aproximate $\infin \approx 135$

In [111]:
import math
import numpy as np

def calculate_M(T, nu, P_hat):
    max_k = round(max(nu*T + 5*math.sqrt(nu*T),20))
    print(max_k)
    total = 0
    for k in range(max_k):
        inner_sum = sum(math.exp(-nu * T + l * math.log(nu * T) - math.lgamma(l + 1)) for l in range(k + 1))
        total += (1 / nu) * (1 - inner_sum) * np.linalg.matrix_power(P_hat, k)
    return total

# Distribution of the first 60 min
occupancy_matrix_60min = calculate_M(60, nu, p_hat)
occupancy_matrix_60min

235


array([[ 3.98884947,  2.16128475, 21.49555701, 10.46371016,  4.32234592,
         3.037422  ,  8.87695254,  5.65387697],
       [ 3.037422  ,  2.18777885, 21.86146471, 10.64387009,  4.39741255,
         3.08599453,  9.03391006,  5.75214602],
       [ 2.78901968,  1.50683213, 22.55878961, 10.81343238,  4.4680635 ,
         2.83759221,  9.18163478,  5.84463454],
       [ 2.84730671,  1.53862506, 21.79787886, 11.02962429,  4.55814347,
         2.89587924,  9.36998379,  5.9625574 ],
       [ 3.07197674,  1.66117235, 21.26491376, 10.66294585,  4.90536078,
         3.12054927,  9.55052909,  5.76255098],
       [ 3.39379183,  1.83670785, 21.46376408, 10.65658726,  4.40271137,
         3.44236436,  9.04498941,  5.75908266],
       [ 2.79873418,  1.51213095, 21.43197115, 10.84946436,  4.48307683,
         2.84730671, 10.21302628,  5.86428835],
       [ 2.76694126,  1.49478935, 21.84701338, 10.7315415 ,  4.43394231,
         2.81551379,  9.56483591,  6.34542134]])

### E[Amount of Time Machine 1 is Blocked]
Machine 1 is blocked in state 3,7, and 8.

Thus, we are interested in the quanitity $m_{43}(60)+m_{47}(60)+m_{48}(60)$

In [112]:
# indexis (4,3), (4,7) and (4,8) are the following 
occupancy_matrix_60min[4-1][3-1], occupancy_matrix_60min[4-1][7-1], occupancy_matrix_60min[4-1][8-1]

(21.797878855208605, 9.369983791057946, 5.962557402341044)

In [113]:
# The sum is the following
occupancy_matrix_60min[4-1][3-1]+occupancy_matrix_60min[4-1][7-1]+occupancy_matrix_60min[4-1][8-1]

37.13042004860759

Thus, in a 60 minute interval, $m_{43}(60)+m_{47}(60)+m_{48}(60)\approx 37.1$ minutes machine 1 is blocked
